### PR TITLE
Update utility network samples to use service geodatabase

### PR DESCRIPTION
**Pull request by vquach2404**
This PR updates utility network samples so that the maps use service geodatabases. Please see `/common-samples/2410` for more info.

### CODE REVIEW

File: arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift:104]:
```suggestion
```
Remove blank lines, linter suggestion.

File: arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift:72]: There were a suggestion by Phil not to use the path component concatenation pattern: https://github.com/Esri/arcgis-runtime-samples-ios/pull/858#discussion_r418925647

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:101]: Same here: https://github.com/Esri/arcgis-runtime-samples-ios/pull/858#discussion_r418925647 we may just directly create an array with hardcoded values.

File: arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift:68]:
```suggestion
        serviceGeodatabase.load { [weak self] _ in
```
Since the error is not checked here, replace it with underscore.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:97]:
```suggestion
        serviceGeodatabase.load { [weak self] error in
```
Also what about moving the error checking to the top of this load method? It seems odd to check error at the end. Though I'm not sure what error would be emitted other than network or load error.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:66]: Hmm it seems to be an odd place to initialize the `utilityNetwork` property... Leave it to Phil's judgement.

**Review by yo1995 — COMMENTED** [review]
The changes look good, and I cannot really spot any difference with ad-hoc testing - which is great! Also, great job for paving the roads for other platforms! 👍 

In addition to the comments below, a few more thoughts

- Since we still don't know when SS7 will go live, it would be painful to try to keep both PRs (#1039 and #1026 ) up-to-date. So once this PR is approved, feel free to put it aside until we revisit it, and solve the merge conflict then. 
- And we may give another approval after testing with the new service URLs.
- It seems without using SS7 URLs, these changes also work without issue... Is that expected? i.e. I can still do the trace in both TUN and PVIT samples.

**Comment by vquach2404** [comment]
> * Since we still don't know when SS7 will go live, it would be painful to try to keep both PRs (#1039 and #1026 ) up-to-date. So once this PR is approved, feel free to put it aside until we revisit it, and solve the merge conflict then.
Yes, I agree. I'll wait until SS7 goes live and update this PR accordingly.
> * It seems without using SS7 URLs, these changes also work without issue... Is that expected? i.e. I can still do the trace in both TUN and PVIT samples.
Oh I see. My hosts file is still edited to accommodate for the SS7 updates. If your hosts file is still edited like mine, you'll have to remove the line in order to test.

**Review by yo1995 — COMMENTED** [review]
LGTM 👍  just one last comment.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (vquach2404) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: That's a good idea! I made a similar change. Please let me know what you think!

**Review by yo1995 — DISMISSED** [review]

File: arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift:30]: It looks like this property isn't even needed as it is only used in `init(coder:)`.

**Comment by vquach2404** [comment]
I've marked this as a draft for now until we can test with the new SS7 updates.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (vquach2404) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:65]: I forwarded this question to Jennifer N. from the utility networks team:

> Yup that is correct, using UN constructor with map parameter will add this UN to the map's UtilityNetworks[]collection. If you use UN constructor with just URL, before loading it, you can add to map.UtilityNetworks property.. but once loaded, you cannot add to a map anymore. This map and UN relationship allows the tables on map to be used by UN - sharing the same instance.

**Comment by vquach2404** [comment]
SS7 updates have gone live! This PR is now ready for reviews. You no longer have to edit your hosts file in order to test. 
Please note that Create load report and Configure subnetwork trace do not need service geodatabases since they don't have maps.

**Review by yo1995 — DISMISSED** [review]

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:65]: Thanks!

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:95]: Methinks this would need to be a property rather than an instance variable as the way it is now, only the completion handler is keeping the geodatabase around, which isn't great. The completion handler should access the geodatabase either through a property of `self` or as an unowned reference.

Comment (vquach2404) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: This change required me to adjust the way the credentials are handled. I also added an additional status message because there were times when the viewpoint took a while to adjust and the status bar said "utility network loaded" which was confusing.

Comment (yo1995) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: I'm a bit confused here - why the credentials handling need to be changed? Could we still apply the credentials to the UN or even to the geodatabase? `serviceGeodatabase.credential = ...`

**Review by yo1995 — COMMENTED** [review]
Please see my comment above.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (vquach2404) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: By making the `serviceGeodatabase` a property, authentication is prompted twice – once for `serviceGeodatabase` and then again for `utilityNetwork`. Using the `AGSAuthenticationManager` allows me to set the credentials once and it will apply the credentials whenever prompted. Otherwise, I'd have to set the same credentials twice for separate instance variables.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:96]: Instead of creating the geodatabase here (and needing it to be mutable and optional), could it be created where the property is declared?

```swift
let serviceGeodatabase = AGSServiceGeodatabase(url: featureServiceURL)
```

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:115]:
```suggestion
```

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (yo1995) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:96]: I think the problem is `featureServiceURL` is a property, which makes the `serviceGeodatabase` cannot access it in the first place.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (yo1995) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: I tried to set the credential on geodatabase and not the UN, and it doesn't prompt me for credentials. This is why I asked...

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:96]: Sounds like `featureServiceURL` should be `static`.

Comment (yo1995) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:96]: ✅  I think this is the right time to pitch the question - should we

- Create all the string/URL literals in a sample as static constants?
- Only create the strings that are required by other properties (like this URL) as static, and other strings as instance property?
- Avoid using static variables, and duplicate the strings when they are required to instantiate an object?

We may agree on a pattern so that we don't have these issues in the future.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [inline arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:61]: Now that creation of the utility network doesn't depend on the map, could we go back to using the `makeMap` method?

Comment (yo1995) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:61]: And since both the `utilityNetwork` and `layers` are properties, probably we can remove the parameters from this method.

File: arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility network/Trace utility network/TraceUtilityNetworkViewController.swift:62]: Should we also update this to `map.utilityNetworks.add(utilityNetwork)`?

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (vquach2404) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: Hm, I could use the `makeMap` method but would have to remove the layers parameter since the service geodatabase must be loaded before I can instantiate the layers array. Please let me know what you think!

**Review by philium — DISMISSED** [review]

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: Yeah, I guess it makes sense to remove the layers parameter. Since the layers need to be added later, it makes sense to do that through the `map` property of the map view, just like you did.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: I don't know that there's a one-size-fits-all pattern we can follow. In this case it made sense for the URL to be static, but sometimes the URL is just used in one place, in which case it doesn't even need to be a property.

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (yo1995) [inline arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift:69]:
```suggestion
```

This line seems to be redundant.

**Review by yo1995 — COMMENTED** [review]
Just one more thing

**Review by yo1995 — APPROVED** [review]

File: arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift
Comment (philium) [reply arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift]: Thank you for catching that!

**Review by philium — APPROVED** [review]

**Review by saratchandrakarumuri — APPROVED** [review]

**Pull request merged**
